File: 02/gpt-5.1/python/solution.py
with open("input.txt") as f:
    input_data = f.read().strip()


def digits_count(n: int) -> int:
    if n < 0:
        n = -n
    count = 1
    while n >= 10:
        n //= 10
        count += 1
    return count


def pow10(exp: int) -> int:
    res = 1
    for _ in range(exp):
        res *= 10
    return res


def sum_invalid_ids(s: str) -> int:
    sum_ids = 0
    for r in s.split(","):
        r = r.strip()
        if not r:
            continue
        parts = [p.strip() for p in r.split("-") if p.strip()]
        if len(parts) != 2:
            continue
        start = int(parts[0])
        end = int(parts[1])
        if end < start:
            start, end = end, start

        min_len = digits_count(start)
        max_len = digits_count(end)
        for length in range(min_len, max_len + 1):
            if length % 2 == 1:
                continue
            half = length // 2
            p10 = pow10(half)
            first = pow10(length - 1)
            last = pow10(length) - 1

            low = max(start, first)
            high = min(end, last)
            if low > high:
                continue

            den = p10 + 1
            a_start = (low + den - 1) // den  # ceil(low/den)
            a_end = high // den  # floor(high/den)
            if a_start > a_end:
                continue

            count = a_end - a_start + 1
            terms_sum = (a_start + a_end) * count // 2
            contrib = terms_sum * den
            sum_ids += contrib
    return sum_ids


part1 = sum_invalid_ids(input_data)
print(f"Day 02 Part 1: {part1}")


def sum_invalid_ids_part2(s: str) -> int:
    sum_ids = 0
    for r in s.split(","):
        r = r.strip()
        if not r:
            continue
        parts = [p.strip() for p in r.split("-") if p.strip()]
        if len(parts) != 2:
            continue
        start = int(parts[0])
        end = int(parts[1])
        if end < start:
            start, end = end, start

        min_len = digits_count(start)
        max_len = digits_count(end)
        for length in range(min_len, max_len + 1):
            for k in range(2, int(length ** 0.5) + 1):
                if length % k != 0:
                    continue

                base_len = length // k
                p10_base = pow10(base_len)
                first = pow10(length - 1)
                last = pow10(length) - 1

                low = max(start, first)
                high = min(end, last)
                if low > high:
                    continue

                geom = 0
                factor = 1
                for _ in range(k):
                    geom += factor
                    factor *= p10_base

                a_start = (low + geom - 1) // geom
                a_end = high // geom
                if a_start > a_end:
                    continue

                count = a_end - a_start + 1
                terms_sum = (a_start + a_end) * count // 2
                contrib = terms_sum * geom
                sum_ids += contrib
    return sum_ids


part2 = sum_invalid_ids_part2(input_data)
print(f"Day 02 Part 2: {part2}")

File: 02/gpt-5.1/typescript/solution.ts
import * as fs from 'fs';

const input = fs.readFileSync('input.txt', 'utf-8').trim();

function digitsCount(n: bigint): number {
	if (n < 0n) n = -n;
	let count = 1;
	while (n >= 10n) {
		n /= 10n;
		count++;
	}
	return count;
}

function pow10(exp: number): bigint {
	let result = 1n;
	for (let i = 0; i < exp; i++) {
		result *= 10n;
	}
	return result;
}

function sumInvalidIds(input: string): bigint {
	let sum = 0n;
	const ranges = input.split(',');

	for (const r0 of ranges) {
		const r = r0.trim();
		if (!r) continue;
		const parts = r.split('-').map(p => p.trim());
		if (parts.length !== 2) continue;

		let start = BigInt(parts[0]);
		let end = BigInt(parts[1]);
		if (end < start) {
			const tmp = start;
			start = end;
			end = tmp;
		}

		const minLen = digitsCount(start);
		const maxLen = digitsCount(end);
		for (let len = minLen; len <= maxLen; len++) {
			if (len % 2 !== 0) continue;

			const half = len / 2;
			const pow = pow10(half);
			const first = pow10(len - 1);
			const last = pow10(len) - 1n;

			let low = start > first ? start : first;
			let high = end < last ? end : last;
			if (low > high) continue;

			const den = pow + 1n;
			let aStart = (low + den - 1n) / den; // ceil(low/den)
			let aEnd = high / den; // floor(high/den)
			if (aStart > aEnd) continue;

			const count = aEnd - aStart + 1n;
			const termsSum = (aStart + aEnd) * count / 2n;
			const contrib = termsSum * den;
			sum += contrib;
		}
	}

	return sum;
}

const part1 = sumInvalidIds(input);
console.log(`Day 02 Part 1: ${part1.toString()}`);

function sumInvalidIdsPart2(input: string): bigint {
	let sum = 0n;
	const ranges = input.split(',');

	for (const r0 of ranges) {
		const r = r0.trim();
		if (!r) continue;
		const parts = r.split('-').map(p => p.trim());
		if (parts.length !== 2) continue;

		let start = BigInt(parts[0]);
		let end = BigInt(parts[1]);
		if (end < start) {
			const tmp = start;
			start = end;
			end = tmp;
		}

		const minLen = digitsCount(start);
		const maxLen = digitsCount(end);
		for (let len = minLen; len <= maxLen; len++) {
			for (let k = 2; k * k <= len; k++) {
				if (len % k !== 0) continue;

				const baseLen = len / k;
				const powBase = pow10(baseLen);
				const first = pow10(len - 1);
				const last = pow10(len) - 1n;

				let low = start > first ? start : first;
				let high = end < last ? end : last;
				if (low > high) continue;

				let geom = 0n;
				let factor = 1n;
				for (let i = 0; i < k; i++) {
					geom += factor;
					factor *= powBase;
				}

				const aStart = (low + geom - 1n) / geom;
				const aEnd = high / geom;
				if (aStart > aEnd) continue;

				const count = aEnd - aStart + 1n;
				const termsSum = (aStart + aEnd) * count / 2n;
				const contrib = termsSum * geom;
				sum += contrib;
			}
		}
	}

	return sum;
}

const part2 = sumInvalidIdsPart2(input);
console.log(`Day 02 Part 2: ${part2.toString()}`);
